src/treelearner/serial_tree_learner.cpp

```diff
@@ -804,51 +774,43 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
   }
 }
 
-
-// this function updates the best split for a leaf
-// it is called only when monotone constraints exist
-void SerialTreeLearner::UpdateBestSplitsFromHistograms(
-    SplitInfo *split, int leaf,
-    const std::vector<int8_t> &is_feature_used_,
-    int num_features_, HistogramPool &histogram_pool_,
-    LearnerState &learner_state) {
-  // the feature histogram is retrieved
-  FeatureHistogram *histogram_array_;
-  histogram_pool_.Get(leaf, &histogram_array_);
+void SerialTreeLearner::RecomputeBestSplitForLeaf(int leaf, SplitInfo* split) {
+  FeatureHistogram* histogram_array_;
+  if (!histogram_pool_.Get(leaf, &histogram_array_)) {
+    Log::Warning(
+        "Get historical Histogram for leaf %d failed, will skip the "
+        "``RecomputeBestSplitForLeaf``",
+        leaf);
+    return;
+  }
   double sum_gradients = split->left_sum_gradient + split->right_sum_gradient;
   double sum_hessians = split->left_sum_hessian + split->right_sum_hessian;
   int num_data = split->left_count + split->right_count;
-  split->gain = kMinScore;
+
+  std::vector<SplitInfo> bests(num_threads_);
+  LeafSplits leaf_splits(num_data);
+  leaf_splits.Init(leaf, sum_gradients, sum_hessians);
 
   OMP_INIT_EX();
-#pragma omp parallel for schedule(static, 1024) if (num_features_ >= 2048)
+// find splits
+#pragma omp parallel for schedule(static)
   for (int feature_index = 0; feature_index < num_features_; ++feature_index) {
     OMP_LOOP_EX_BEGIN();
-    // splits are computed for the feature that are to be used
-    if (!histogram_array_[feature_index].is_splittable()) {
+    if (!is_feature_used_[feature_index]) {
       continue;
     }
-
-    // loop through the features to find the best one just like in the
-    // FindBestSplitsFromHistograms function
-    int real_fidx = learner_state.train_data_->RealFeatureIndex(feature_index);
-    LeafSplits *leaf_splits = new LeafSplits(0);
-    leaf_splits->Init(leaf, sum_gradients,
-                     sum_hessians);
-
-    // best split is updated
+    const int tid = omp_get_thread_num();
+    int real_fidx = train_data_->RealFeatureIndex(feature_index);
     ComputeBestSplitForFeature(
-        histogram_array_,
-        feature_index,
-        real_fidx,
-        is_feature_used_[feature_index], // FIXME is this the right parameter to pass here?
-        num_data,
-        leaf_splits,
-        split);
+        smaller_leaf_histogram_array_, feature_index, real_fidx,
+        true,  // fixme: this cannot work with colsample_bynode, but do we really need to compute all features? why not just the used features of `split->feature`?
```

Comment:
Do we really need to scan all features?
If yes, we can pre-check the `colsample_bynode` in config, and don't enable them at the same time.

Reply:
We don't need to scan all the features, we just need to scan the ones we want to split on. sorry I didn't do that from the start, I didn't know this parameter existed to be honest.

Reply:
@guolinke can you advise what to use instead of true here then? Thanks

Reply:
double confirm: For one leaf, we only need to calculate the gain for one feature, which is the feature stored in `splitInfo`, right?

Reply:
No, what I meant is that we need to compute the split for all feature that we would normally use. The reason for that is that say the split info is on one feature at some point but the constraints are updated. Then, this feature may not be the optimal feature to split on anymore, so we need to look at all available features all over again. 

Does that make sense? Thanks

Reply:
okay, got you.
This `true` stands for, whether this feature is chosen in the current leaf or not, refer to the parameter `colsample_bynode`. 
So, to work with `colsample_bynode`, we may need to store the historical chosen features for all leaves. I think this is a large cost.
A simple solution is to disable one of them when both `colsample_bynode` and `MC` is set.

Reply:
I think the way to look at it is that basically, all the work to find the best split that was done before the constraints are updated is useless once the constraints are updated. So the code needs to look at all the features it looked at the first time, and determine once again where the best split is, given the new constraints.

Reply:
Yes storing the the historical features chosen may be the best way to do that. Otherwise, maybe it would possible to re sample each time, but that does not seem like such a great idea. I will add a commit to disable one when the other is set as you said, and that can always be improved later! Thanks

Reply:
yeah, refer to https://github.com/CharlesAuguste/LightGBM/pull/2/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h#diff-19606143d392a547b85ff5957e68ab8eR314-R318, you can disable in config checking.